daemon/libnetwork/internal/nftables/nftables_linux.go

```diff
@@ -670,108 +861,180 @@ func (v VMapRef) DeleteElement(ctx context.Context, key string) error {
+	return true, nil
 }
 
-// DeleteElement deletes an element from the set. It is an error to delete an
-// element that is not in the set.
-func (s SetRef) DeleteElement(ctx context.Context, element string) error {
-	if _, ok := s.s.Elements[element]; !ok {
-		return fmt.Errorf("set does not contain element %q", element)
+func (se SetElement) delete(ctx context.Context, t *table) (bool, error) {
+	s := t.Sets[se.SetName]
+	if s == nil {
+		return false, fmt.Errorf("cannot delete from set '%s', it does not exist", se.SetName)
 	}
-	delete(s.s.Elements, element)
-	s.s.DeletedElements[element] = struct{}{}
+	if _, ok := s.Elements[se.Element]; !ok {
+		return false, fmt.Errorf("cannot delete '%s' from set '%s', it does not exist", se.Element, s.Name)
+	}
+	delete(s.Elements, se.Element)
+	delete(s.AddedElements, se.Element)
+	s.DeletedElements[se.Element] = struct{}{}
 	log.G(ctx).WithFields(log.Fields{
-		"family":  s.s.table.Family,
-		"table":   s.s.table.Name,
-		"set":     s.s.Name,
-		"element": element,
-	}).Debug("nftables: deleted set element")
-	return nil
+		"family":  t.Family,
+		"table":   t.Name,
+		"set":     s.Name,
+		"element": se.Element,
+	}).Debug("nftables: added set element")
+	return true, nil
 }
 
 // ////////////////////////////
 // Internal
 
+func (t *table) deleteChain(ctx context.Context, name string) (bool, error) {
+	c := t.Chains[name]
+	if c == nil {
+		return false, fmt.Errorf("cannot delete chain '%s', it does not exist", name)
+	}
+	if len(c.ruleGroups) != 0 {
+		return false, fmt.Errorf("cannot delete chain '%s', it is not empty", name)
+	}
+	delete(t.Chains, name)
+	t.DeleteCommands = append(t.DeleteCommands,
+		fmt.Sprintf("delete chain %s %s %s", t.Family, t.Name, name))
+	log.G(ctx).WithFields(log.Fields{
+		"family": t.Family,
+		"table":  t.Name,
+		"chain":  name,
+	}).Debug("nftables: deleted chain")
+	return true, nil
+}
+
+type command struct {
+	obj        Obj
+	callerFile string
+	callerLine int
+	delete     bool
+}
+
+func (c command) apply(ctx context.Context, t *table) (bool, error) {
+	if c.delete {
+		return c.obj.delete(ctx, t)
+	}
+	return c.obj.create(ctx, t)
+}
+
+func (c command) rollback(ctx context.Context, t *table) (bool, error) {
+	if c.delete {
+		return c.obj.create(ctx, t)
+	}
+	return c.obj.delete(ctx, t)
+}
+
 func (t *table) updatesApplied() {
-	t.DeleteChainCommands = t.DeleteChainCommands[:0]
+	t.DeleteCommands = t.DeleteCommands[:0]
 	for _, c := range t.Chains {
-		c.Dirty = false
+		c.MustFlush = false
 	}
 	for _, m := range t.VMaps {
-		m.Dirty = false
 		m.AddedElements = map[string]string{}
-		m.DeletedElements = map[string]struct{}{}
+		m.DeletedElements = map[string]string{}
+		m.MustFlush = false
 	}
 	for _, s := range t.Sets {
-		s.Dirty = false
 		s.AddedElements = map[string]struct{}{}
 		s.DeletedElements = map[string]struct{}{}
+		s.MustFlush = false
 	}
-	t.Dirty = false
+	t.MustFlush = false
 }
 
 /* Can't make text/template range over this, not sure why ...
```

Comment:
range-over-func support in text/template [was added in Go 1.24.](https://go.dev/doc/go1.24#texttemplatepkgtexttemplate) That will need to wait until we bump the minimum Go version.